### PR TITLE
Feature / Audit Logging

### PR DESCRIPTION
**Pull request by bryan-robitaille**
# *Has infrastructure dependency*
cds-snc/forms-terraform#336
# Summary | Résumé

This PR adds the ability to log events of importance across the application to be stored separately from the application logs.

**Addition of `logEvent` function that takes the user's ID, the type and object ID being acted upon, the action, and a description.**

- The SQS message queue instance and Queue URL are initialized on the first call and instances are reused.
- The `logEvent` function is asynchronous and fails silently to ensure there are no 'unhandled exceptions'.
- If the `logEvent` function fails the log events are diverted into the applications logs (CloudWatch) to be retrieved at a later date.

**Other application changes**:
- Dropped the API User logs table as they will eventually be included in the logEvents
- Dropped the Admin Log tables as they are now directed to logEvents
- Upgraded Next-Auth to the latest version.  `unstable_getServerSession` is now `getServerSession`
- Temporary Token , Bearer Token, and API User endpoints are disabled until the functionality can be reviewed and modified to use the new User models.
- Added the user's ID to the `ability` object.  Instead of adding another argument to the majority of lower level functions it was more efficient to add the userID to the `ability` object as it was already introduced in all the lower level functions. 
- Added a check to `notify-callback` to use an environment variable first, if it exists, for the queue URL before querying the AWS infrastructure.

**Jest Testing changes**:
- Added better support to ensure common required environment variables are injected
- Removed `@types/jest` and replaced with `@jest/globals` due to not all Jest types being included in 3rd party supported package.   Functions not included but of interest include `jest.Replaced<>` and `jest.replaceProperty`
- 

# Test instructions | Instructions pour tester la modification

## Without Localstack setup from cds-snc/forms-terraform#336
Perform various actions like logging in, changing privileges, and creating and modifying forms.  The log events should be sent to your applications console in the terminal.  A `debug` message should contain the log event information.

## With Localstack setup from cds-snc/forms-terraform#336
Comment out this line in [`setupFiles.js`](https://github.com/cds-snc/platform-forms-client/blob/fb3a40386e3559c8d5d03c0f3c0b9af1ed8d7b38/__utils__/setupTests.ts#L26)
This will allow Jest tests to actually create events through localstack.
1. `yarn test`
2. In Infrastructure repo run `./invoke_audit_logs` under `aws/app/lambdas`
3. Do a scan on the AuditLogs table in DynamoDB from CLI `aws dynamodb scan --table-name AuditLogs --endpoint http://localhost:4566` or  open Dynamodb connection from NoSQLWorkbench
4. Table should be populated with Log Events

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

The functionality to extract Audit Logs from Cloudwatch into Dynamodb in event of a catastrophic failure has yet to be implemented.

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [x] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [ ] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [ ] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [ ] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

File: __tests__/api/flags.test.ts [hunk 1]
```diff
@@ -2,23 +2,26 @@
...
 const redis = new Redis();
 
 jest.mock("@lib/integration/redisConnector", () => ({
   getRedisInstance: jest.fn(() => redis),
 }));
 
 jest.mock("next-auth/next");
+jest.mock("@lib/auditLogs");
 
 //Needed in the typescript version of the test so types are inferred correclty
```
Comment (timarney) [inline __tests__/api/flags.test.ts:22]: typo - correctly

File: __tests__/api/id/form/apiUsers.test.ts [hunk 2]
```diff
@@ -5,21 +5,19 @@
...
-import { logAdminActivity } from "@lib/adminLogs";
 import { prismaMock } from "@jestUtils";
 import { Prisma } from "@prisma/client";
 import { Session } from "next-auth";
 import { Base, ManageForms, getUserPrivileges } from "__utils__/permissions";
 
 jest.mock("next-auth/next");
-jest.mock("@lib/adminLogs");
 
 //Needed in the typescript version of the test so types are inferred correclty
```
Comment (timarney) [inline __tests__/api/id/form/apiUsers.test.ts:17]: typo - correctly

File: __tests__/api/id/form/bearer.test.ts [hunk 3]
```diff
@@ -4,24 +4,22 @@
...
 import jwt from "jsonwebtoken";
-import { logAdminActivity } from "@lib/adminLogs";
 import { prismaMock } from "@jestUtils";
 import { Prisma } from "@prisma/client";
 import { Session } from "next-auth";
 
 jest.mock("next-auth/next");
-jest.mock("@lib/adminLogs");
 
 //Needed in the typescript version of the test so types are inferred correclty
```
Comment (timarney) [inline __tests__/api/id/form/bearer.test.ts:17]: type - correctly

**Review by craigzour — APPROVED** [review]
LGTM! Lots of improvements. Great work :)
I think https://github.com/cds-snc/platform-forms-client/issues/740 is not relevant anymore

Will run local testing and approve the terraform PR when I am done.

File: __tests__/api/id/form/apiUsers.test.ts [hunk 4]
```diff
@@ -5,21 +5,19 @@
...
-const mockGetSession = jest.mocked(unstable_getServerSession, { shallow: true });
+const mockGetSession = jest.mocked(getServerSession, { shallow: true });
 
-describe("/id/[forms]/owners", () => {
+describe.skip("/id/[forms]/owners", () => {
```
Comment (thiessenp-cds) [inline __tests__/api/id/form/apiUsers.test.ts:20]: This skip is intentional? Suspect it is but double checking

**Review by craigzour — APPROVED** [review]

File: __tests__/api/id/form/apiUsers.test.ts (hunk 4 above)
Comment (bryan-robitaille) [reply __tests__/api/id/form/apiUsers.test.ts:20]: It is because we've disabled the API endpoint by passing in [] to `allowedMethods`.

Comment (thiessenp-cds) [reply __tests__/api/id/form/apiUsers.test.ts:20]: Right, I remember seeing that for a POST endpoint but not others. Need more coffee. Good - will resolve this :)

**Review by thiessenp-cds — COMMENTED** [review]
LGTM! Impressive work, especially the unit tests.
Note: reminder for a `yarn install` or you'll be surprised like me with a lot of failing tests on the first run :)

**Review by craigzour — APPROVED** [review]

**Pull request merged**